src/rpp/rpp/operators/buffer.hpp

```diff
@@ -0,0 +1,83 @@
+//                  ReactivePlusPlus library
+//
+//          Copyright Aleksey Loginov 2022 - present.
+//                    TC Wang 2022 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/observables/constraints.hpp>
+#include <rpp/operators/fwd/buffer.hpp>
+#include <rpp/subscribers/constraints.hpp>
+#include <rpp/sources/create.hpp>
+
+IMPLEMENTATION_FILE(buffer_tag);
+
+namespace rpp::details
+{
+
+/// A non-copyable class that provides a copyable on_next for the subscriber and
+/// allows copies of on_next(s) to share the same states.
+template<constraint::decayed_type UpstreamType, constraint::decayed_type DownstreamType>
+struct buffer_on_next : public std::enable_shared_from_this<buffer_on_next<UpstreamType, DownstreamType>>
+{
+    using DownstreamValueType = UpstreamType;
+
+    explicit buffer_on_next(const size_t& count) : max_(count) {};
+
+    // Copy & move constructors
+    buffer_on_next(const buffer_on_next& other) = delete;
+    buffer_on_next(buffer_on_next&&) noexcept = default;
+
+    // Copy & move assignment operators
+    buffer_on_next& operator=(const buffer_on_next&) = delete;
+    buffer_on_next& operator=(buffer_on_next&&) noexcept = default;
+
+    /// \return a on_next function for subscriber, that is copyable and all the
+    /// copies share the same state.
+    auto get_on_next() {
+        return [shared_this = this->shared_from_this()](
+                const UpstreamType& value,
+                const constraint::subscriber_of_type<DownstreamType> auto& subscriber)
+        {
+            shared_this->buckets_.push_back(value);
+            if (shared_this->buckets_.size() == shared_this->max_)
+            {
+                auto copy = std::vector<DownstreamValueType>(shared_this->buckets_);
+                shared_this->buckets_.clear();
+                subscriber.on_next(std::move(copy));
+            }
+        };
+    }
+
+private:
+    const size_t max_;
+    std::vector<DownstreamValueType> buckets_;
+};
+
+template<constraint::decayed_type UpstreamType, constraint::decayed_type DownstreamType>
+struct buffer_impl
+{
+    const size_t count;
+
+    auto operator()(auto&& subscriber) const
```

Comment:
Add check over subscriber to avoid some template issues in future
```suggestion
   
    template<constraint::subscriber_of_type<DownstreamType> TSub>
    auto operator()(TSub&& subscriber) const
```

Reply:
Sounds good. For my education, what are those potential issues?

Reply:
Your operator expects to obtain a subscriber of some specific type (let's say of `int`), but in case of obtaining subscriber of another type (let's say of `string`) you would accept some non-understandable issues related to inability to pass `string` where `int` expected. I think, it is better to have error like `expected subscriber of ints but obtained subscriber of strings`. Currently it is hard to obtain such an error, but anyway it improves overall readability of code

Reply:
Thanks for the explanation, and I'll update the code.